File: Components/Strata.Foundation.Core/main/typescript/Utility/IOptional.ts
import {ISupplier} from "./ISupplier";
import {IConsumerOrLambda} from "./LambdaConsumer";
import {IFunctionOrLambda} from "./LambdaFunction";
import {ISupplierOrLambda} from "./LambdaSupplier";
import {IRunnableOrLambda} from "./LambdaRunnable";
import {IPredicateOrLambda} from "./LambdaPredicate";

export
interface IOptional<T>
    extends ISupplier<T>
{
    ifPresent(consumer: IConsumerOrLambda<T>): void;

    ifPresentOrElse<U>(
        presentAction: IFunctionOrLambda<T,U>,
        emptyAction: ISupplierOrLambda<U>): U;

    ifPresentOrElseNoReturn(
        consumer: IConsumerOrLambda<T>,
        action: IRunnableOrLambda): void;

    or(supplier: ISupplierOrLambda<IOptional<T>>): IOptional<T>;

    orElse(alternative: T): T;

    orElseGet(supplier: ISupplierOrLambda<T>): T;

    orElseThrow(error: Error): T;

    orElseGetThrow(supplier: ISupplierOrLambda<Error>): T;

    filter(predicate: IPredicateOrLambda<T>): IOptional<T>;

    map<U>(mapper: IFunctionOrLambda<T,U>): IOptional<U>;

    flatMap<U>(mapper: IFunctionOrLambda<T,IOptional<U>>): IOptional<U>;

    isPresent(): boolean;

    isEmpty(): boolean;

}

File: Components/Strata.Foundation.Core/main/java/strata/foundation/core/utility/IOptional.java
//////////////////////////////////////////////////////////////////////////////
// IOptional.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import java.util.NoSuchElementException;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public
interface IOptional<T>
    extends Supplier<T>
{
    void
    ifPresent(Consumer<T> consumer);

    <U> U
    ifPresentOrElse(Function<T,U> present,Supplier<U> notPresent);

    <U,E extends RuntimeException> U
    ifPresentOrThrow(Function<T,U> present,E exception);

    void
    ifPresentOrElseNoReturn(Consumer<T> present,Runnable notPresent);

    <E extends RuntimeException> void
    ifPresentOrThrowNoReturn(Consumer<T> present,E exception) throws E;

    void
    ifNotPresent(Runnable notPresent);

    IOptional<T>
    filter(Predicate<T> predicate);

    <U> IOptional<U>
    map(Function<T,U> mapper);

    <U,O extends IOptional<U>> IOptional<U>
    flatMap(Function<T,O> mapper);

    <O extends IOptional<T>> IOptional<T>
    or(Supplier<O> supplier);

    T
    orElse(T other);

    T
    orElseGet(Supplier<T> other);

    T
    orElseThrow() throws NoSuchElementException;

    <E extends RuntimeException> T
    orElseThrow(Supplier<E> exceptionSupplier) throws E;

    boolean
    isPresent();

    boolean
    isEmpty();

    Optional<T>
    toOptional();

}

//////////////////////////////////////////////////////////////////////////////